### DISCUSSION TITLE
Template scaffolding

### DISCUSSION THREAD

**Discussion by erikyo** [wp-blocks/wp-blocks#3] [Ideas]
I must say that I liked the idea after all, but I am not a big fan of moving the whole template into a subfolder so I created this package hacking the @johnhooks stuff

https://github.com/erikyo/create-pkg-cli

in short we can publish this as a npm package and then on postinstall run the script (eventually with npx)

@johnhooks what do you think? might be a good way to do this tasks without having to copy unnecessary files and without using subfolders

**Comment by johnhooks** [comment]
Absolutely, I'm still going to work on the more complex idea. But we can move forward on both and see which we like better.

Are you up for making the main repo we fork from under the `wp-block` org? You should be able to create a repo there. 

Also, I the code I did was kinda of hacky, I was planning to move the code under `lib` to be all TS, move it to a `src` directory and we can build the output to the `build`. What do you think of that?

I'm almost done with a `Config` type that would fit the block project better, I'll try to get that published soon, it might be handy.

**Reply by johnhooks** [reply]
Also, I think something like `@wp-blocks/tsconfig` could save us a lot of time. I'll work on that.

**Reply by johnhooks** [reply]
Oh I see it's already there! I also published the first npm package for the project!

https://www.npmjs.com/package/@wp-blocks/tsconfig

**Reply by erikyo** [reply]
moved the create-pkg-cli into the org, the only thing is that I think it would be better to delete it and republish it since 99% of the code is yours and I didn't want to steal your work :)

anyway it is intended to be published into the npm registry, in this way we can call this script on git clone

**Reply by johnhooks** [reply]
@erikyo don't worry about that. I'm hoping we replace it all anyway 😉

**Comment by johnhooks** [comment]
@erikyo I still don't understand 100% how you intend to use this. Why would it need to be listed as a dependency? My original intent is for it to be used like

```sh
# scaffold a full plugin project
npm create @wp-blocks/plugin

# scaffold just a block 
npm create @wp-blocks/block 
```

And it uses the templates in the subfolder to generate the files for the package, it's different in concept from using `git clone`. The the output is only all the generated files.

**Reply by erikyo** [reply]
I would to create a a npx command that scaffolds something from a template, maybe a plugin or a block. Maybe in this case could fit better `@wp-blocks/scaffold` with some options like '-block' or '-plugin' etc

And I agree, the ideal thing would be to have a minimum output with only the strictly necessary files and folders, at least in the case of the block option.

**Comment by johnhooks** [comment]
@erikyo Do you know about npm initializers? If you use the word `create` as the first world of the unscoped package name npm will treat it differently. Read the [npm-init Description section](https://docs.npmjs.com/cli/v6/commands/npm-init#description)

Example: A package named `@wp-blocks/create-block` can be called like so to scaffold a project.

```sh
npm create @wp-blocks/block  
```

It adds the `create` into the name, and searches for a package named `@wp-blocks/create-block`, downloads it, then npm will call the script configured under `"bin"` in the `package.json` of the `@wp-blocks/create-block`. The package itself isn't used as a template, but the "create" package can build out the template and then it's over, nothing of the "create" package is retained in the project folder, other than what it creates.

**Reply by johnhooks** [reply]
This is what `@bitmachina/create-pkg` does if used like:

```
npm create @bitmachina/pkg
```

**Reply by erikyo** [reply]
Ah ok, yes, it was exactly the way I wanted to use the package, so need for npx? but isn't the difference that for npx you don't have to have anything previously installed and it is a command that you only run once? I have to do some testing to see which way is more practical to use, but if it can be done without npx as you say then it is better
